### ISSUE TITLE
Include dependencies in the hash

### ISSUE THREAD

**Issue by int-index** [kowainik/smuggler#39]
Let's say I have three modules: `A`, `B`, and `C`:

```
module A (X) where
data X
```

```
module B (Y) where
data Y
```

```
module C where
import A
import B
data Z = MkZ X Y
```

Then I change `B`:

```
module B (X, Y) where
import A (X)
data Y
```

Now `import A` in `C` is redundant, because `B` reexports `X`. However, `C` itself hasn't changed.

This means that using `hash-store` to decide whether recompilation is needed is not sufficient, because redundancy of imports is not a property of a particular module, but a property of this module and all its dependencies.

**Comment by chshersh** [comment]
This seems a little hard to solve in efficient way. What can be done as a temporal workaround: not use `hash-store` at all, find unused imports every time and check whether there're any of them. Not sure how slow it will become though.

**Comment by int-index** [comment]
> a little hard to solve in efficient way

I think it can be solved efficiently by including the hashes of all imported modules in the hash of each module. For instance, if we have

```
module A where

import B
import C
```

then we compute `hashOf(module A)` as `hashOf(A.hs file) + hashOf(module B) + hashOf (module C)`.

**Comment by chshersh** [comment]
@int-index But some modules are from other libraries. I don't have access to their source code. So I need somehow to detect which modules are from project itself. And I need to calculate graph of dependencies and transitive closure. `ghcid` by default compiles modules in parallel. If module `A` depends on `B` I don't have guarantees that module `A` will be processed only after `B` changes. A lot of complicated logic will be added to support rather unpopular case.

It's just looks like over-engineering for simple problem. I don't claim that this problem shouldn't be solved. I just want simpler solution :)

There's GSoC project regarding caching output of `ghc` compilation. After it's done, I don't need to call `findImportUsage` function manually. GHC does this anyway, so I can just reuse compilation output. And the problem will be solved without performance regression.

**Comment by chshersh** [comment]
I want to point out that the last example is not exactly correct:

```haskell
module C where
import A
import B
data Z = MkZ X Y
```

GHC reports unused imports in this case only when `import B` goes before `import A`. The behaviour probably is different in recently released GHC-8.8 (I didn't check, but I saw some improvements in the algorithm regarding finding unused imports). `smuggler` doesn't have a goal to be smarter than GHC. I made some changes to `smuggler` in PR #67, and now it works correctly in this case. It even works with on-the-fly mode with `ghcid` if you change `module B`: in that case `module C` is automatically refactored for you (but only if imports are specified in `module B` in proper order).
